org.eclipse.jdt.ls.tests/src/org/eclipse/jdt/ls/core/internal/correction/UnresolvedTypesQuickFixTest.java

```diff
@@ -1400,4 +1403,30 @@ public void testDontImportTestClassesInMainCode() throws Exception {
 		assertCodeActionNotExists(cu1, "Import 'Tests' (pt)");
 	}
 
+	@Test
+	public void testTypeInSealedTypeDeclaration() throws Exception {
+		Map<String, String> options15 = new HashMap<>();
+		JavaModelUtil.setComplianceOptions(options15, JavaCore.VERSION_15);
+		options15.put(JavaCore.COMPILER_PB_ENABLE_PREVIEW_FEATURES, JavaCore.ENABLED);
+		options15.put(JavaCore.COMPILER_PB_REPORT_PREVIEW_FEATURES, JavaCore.IGNORE);
+		fJProject1.setOptions(options15);
+
+		IPackageFragment pack1 = fSourceFolder.createPackageFragment("test1", false, null);
+		StringBuilder buf = new StringBuilder();
+		buf.append("package test1;\n");
+		buf.append("public sealed interface E permits F {\n");
```

Comment:
I'll have another look at this. The initial problem I had was that the "add unimplemented methods" quick fix would fail because the ASTRewrite complained if the underlying CU didn't yet exist (which is the case here for F). This is why a lot of the code uses literal values to generate the code (eg. StubUtility* helpers). I could try calling StubUtility2Core.createImplementationStubCore(..) directly once we get the unimplemented methods but even that seems to require an ASTRewrite.

Reply:
So given this is a more global issue with generated types, we'll address this in a separate issue.